upstream/upopenfaas/provision.go

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) 2019-present Mattermost, Inc. All Rights Reserved.
+// See License for license information.
+
+package upopenfaas
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/openfaas/faas-cli/stack"
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v3"
+
+	"github.com/mattermost/mattermost-plugin-apps/apps"
+	"github.com/mattermost/mattermost-plugin-apps/upstream"
+	"github.com/mattermost/mattermost-plugin-apps/utils"
+)
+
+// ProvisionApp creates OpenFaaS functions from an app bundle, as declared by
+// the app's manifest.
+func ProvisionApp(bundlePath string, log utils.Logger, shouldUpdate bool, gateway, prefix string) (*apps.Manifest, error) {
+	m, dir, err := upstream.GetAppBundle(bundlePath, log)
+	if err != nil {
+		return nil, err
+	}
+	defer os.RemoveAll(dir)
+
+	yamlFile := filepath.Join(dir, ManifestYaml)
+	parsedServices, err := stack.ParseYAMLFile(yamlFile, "", "", false)
+	if err != nil {
+		return nil, err
+	}
+	parsedServices.Provider.GatewayURL = gateway
+
+	newFunctions := map[string]stack.Function{}
+	for name, origF := range parsedServices.Functions {
+		f := origF
+		if f.Environment == nil {
+			f.Environment = map[string]string{}
+		}
+		f.Name = FunctionName(m.AppID, m.Version, name)
+		if !strings.HasSuffix(prefix, "/") {
+			prefix += "/"
+		}
+		f.Image = prefix + f.Image
+		newFunctions[f.Name] = f
+	}
+	parsedServices.Functions = newFunctions
+
+	yamlData, err := yaml.Marshal(parsedServices)
+	if err != nil {
+		return nil, err
+	}
+	err = os.WriteFile(yamlFile, yamlData, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	// Provision functions.
+	faascliPath, err := exec.LookPath("faas-cli")
```

Comment:
Instead of relying on the CLI tool being installed, should we execute the command directly: https://github.com/openfaas/faas-cli/blob/a230119be00f5fb832f0c5e92b39576e2fb1c91e/commands/faas.go#L72

Reply:
Not sure about that... suspicious about `faasCmd.SilenceErrors = true` and `checkAndSetDefaultYaml()`

Is there anything wrong (corner case?) with this implementation? If we don't have faas-cli configured, I am not sure all the authentication stuff would work. I think it's an effective requirement.

Reply:
My comment was to hard. I'm just wondering how the requirements on the CLI tool will interact with the plan to move `appsctl` to the `/apps` (https://github.com/mattermost/mattermost-plugin-apps/pull/260#discussion_r720355070).

From reading the docs at https://github.com/openfaas/faas-cli it's unclear anyway how authentication to openfass works. Is there a token used somwhereß

Reply:
It'd require further investigation. 1/5 agree that directly using APIs with a token is the preferred way, future work item?

Reply:
Surely, this doesn't block the PR.